### PR TITLE
[SAMZA-2772] Manage Jekyll dependencies with Gradle

### PR DESCRIPTION
**Pull request by li-afaris**
# Summary
This change moves Jekyll dependency management into Gradle. Contributors will no longer be required to install ruby & Jekyll by hand.

# Changes
* Gemfiles no longer manage dependencies.  This is done with the jruby-gradle-plugin Gradle plugin.  
     * rubygems.org is still the source of Ruby libraries
* The locally install MRI Ruby interpreter is replaced with JRuby.  
     * This allows the Ruby interpreter to be a part of the Samza project.
* Jekyll plugin changes
     * Rouge replaces pygments.rb. Rouge is native Ruby and does not have a python requirement.
     * Kramdown replaces Redcarpet because Redcarpet does not work with JRuby.
     * Both Kramdown & Rouge can run inside the JRuby JVM.
* buildscript.gradle
     * jcenter.bintray.com is no longer active.
     * plugins.gradle.org has the required plugins

# Testing

Testing was done with JDK 1.8 and on Linux.  I visually confirmed that content generated with JRuby looks like content generated by MRI Ruby by running two Jekyll instances at the same time.  Both Jekyll instances threw the same errors for missing content at `/learn/documentation/latest`.   I'm open to suggestions for improving confidence relating to the Jekyll plugin changes.

I also ran `./gradlew test` which built successfully.

### CODE REVIEW

**Comment by li-afaris** [comment]
I'm not sure why the last commit (ccb932d) failed testing when the previous one (651a1bc) worked.  You can see the diff by clicking the ccb932d link above.

Is the samza-kafka_2.12:test task stable? I ran `./gradlew clean build test` locally and the build was successful.  Let me know if I need to trigger another build with an empty commit or if it's something else.  Thanks

**Comment by li-afaris** [comment]
> I'm open to suggestions for improving confidence relating to the Jekyll plugin changes.

One thing I thought of was to list & compare the generated file names in _site.  I found some of the javadoc files were missing with the jruby generated site.  I think I need to revisit the Jekyll plugin changes.

**Comment by amf3** [comment]
I found the following differences after generating documents with Jekyll using both MRI Ruby & Jruby.  Input was generated by running find on the _site directory and saving the output to a sorted text file.

```
$ sdiff -Ws -w 220 ~/mri_ruby_docs ~/jruby_docs  
./Gemfile                                                                                                    <
./Gemfile.lock                                                                                            <
./index.md.bak                                                                                          <
./meetups/aug-2016.md                                                                          <
```

The missing Gemfile, Gemfile.lock, & index.md.bak files I'm not concerned with as they were removed as part of this change.  The meetups/aug-2016.md file was a surprise.  FWIW: it's not referenced in the MRI Ruby build or at https://samza.apache.org/meetups so I'm not concerned about it right now.

**Comment by li-afaris** [comment]
FYI: Removing the Jekyll incremental build flag resolved the file name differences I found in the previous comment.

File: docs/meetups/index.html [hunk 1]
```diff
@@ -21,7 +21,7 @@
 -->
 
 <div class="meetups" data-plugin="paginate" data-max-posts="2" data-previous=".paginate-previous--meetup" data-next=".paginate-next--meetup" data-pagination=".pagination--meetup">
-    {% assign sorted = (site.meetups | sort: 'date') | reverse %}
+    {% assign sorted = site.meetups | sort: 'date' | reverse %}
```
Comment (mynameborat) [inline docs/meetups/index.html:24]: why do we need these changes?

File: docs/talks/index.html [hunk 2]
```diff
@@ -21,7 +21,7 @@
 -->
 
 <div class="talks" data-plugin="paginate" data-max-posts="2" data-previous=".paginate-previous--talk" data-next=".paginate-next--talk" data-pagination=".pagination--talk">
-    {% assign sorted = (site.talks | sort: 'date') | reverse %}
+    {% assign sorted = site.talks | sort: 'date' | reverse %}
```
Comment (mynameborat) [inline docs/talks/index.html:24]: same as above. can we exclude these changes if they are not part of the original issue this RB is solving?

File: docs/_blog/index.md [hunk 3]
```diff
@@ -22,7 +22,7 @@ exclude_from_loop: true
 
 <div class="news_cards">
 
-  {% assign sorted = (site.blog | sort: 'date') | reverse %}
+  {% assign sorted = site.blog | sort: 'date' | reverse %}
```
Comment (mynameborat) [inline docs/_blog/index.md:25]: revert unless its related to this PR

**Review by mynameborat — COMMENTED** [review]
Thanks for fixing this Adam! It looks a lot cleaner now.

File: docs/talks/index.html (hunk 2 above)
Comment (li-afaris) [reply docs/talks/index.html:24]: Yes this is due to the change from redcarpet to kramdown

File: docs/_blog/index.md (hunk 3 above)
Comment (li-afaris) [reply docs/_blog/index.md:25]: Yes this is due to the change from redcarpet to kramdown

File: docs/meetups/index.html (hunk 1 above)
Comment (li-afaris) [reply docs/meetups/index.html:24]: Yes this is due to the change from redcarpet to kramdown

**Review by ajothomas — APPROVED** [review]
Looks good. Tested in on Mac and Linux (centos7). Thanks for the patch.

**Review by ajothomas — CHANGES_REQUESTED** [review]
Please update `bin/publish-site.sh` to use the gradle task instead of `bundle exec jekyll build`

**Comment by li-afaris** [comment]
Good catch.  I updated publish-site.sh and tested the best that I could as I don't have SVN write access.  Let me know if it doesn't work correctly.  Thanks

**Review by ajothomas — APPROVED** [review]
Looks good. Thanks for the patch.

**Pull request merged**